### PR TITLE
Add support for partial_apply and convert_escape_to_noescape

### PR DESCRIPTION
**Pull request by apaszke-g**

### CODE REVIEW

File: Sources/SIL/SILParser.swift [hunk 1]
```diff
@@ -182,11 +189,23 @@ class SILParser: Parser {
             let value = try parseInt()
             return .integerLiteral(type, value)
         case "load":
+            let _ = skip("[copy]")
+            let _ = skip("[take]")
+            let _ = skip("[trivial]")
```
Comment (burmako) [inline Sources/SIL/SILParser.swift]: Let's have these as fields on `load` (and add a test that uses all these flags)?

File: Sources/SIL/SILParser.swift [hunk 2]
```diff
@@ -199,6 +218,7 @@ class SILParser: Parser {
         case "store":
             let value = try parseValue()
             try take("to")
+            let _ = skip("[init]") // Used in ownership SSA
             let _ = skip("[trivial]") // Used in ownership SSA
```
Comment (burmako) [inline Sources/SIL/SILParser.swift]: Likewise, let's add these flags to fields of `store`.

File: Sources/SIL/SILParser.swift [hunk 3]
```diff
@@ -610,7 +630,7 @@ class SILParser: Parser {
     func parseTypeName() throws -> String {
         let start = position
         // TODO(#14): Make name parsing more thorough.
-        let name = take(while: { $0.isLetter || $0.isNumber || $0 == "_" })
+        let name = take(while: { $0.isLetter || $0.isNumber || "_[]".contains($0) })
```
Comment (burmako) [inline Sources/SIL/SILParser.swift]: When is this necessary? Let's add a test?

File: Sources/SIL/SILPrinter.swift [hunk 4]
```diff
@@ -163,6 +170,15 @@ class SILPrinter: Printer {
...
+            print(when: calleeGuaranteed, "[callee_guaranteed] ")
+            print(when: onStack, "[on_stack] ")
+            print(value)
+            print(whenEmpty: false, "<", substitutions, ", ", ">") { naked($0) }
+            print("(", arguments, ", ", ")") { print($0) }
+            print(" : ")
+            print(type)
         case let .pointerToAddress(operand, strict, type):
             print("pointer_to_address ")
             print(operand)
```
Comment (burmako) [inline Sources/SIL/SILPrinter.swift]: Am I missing something or this PR doesn't have instruction tests?

File: Sources/SIL/SILParser.swift (hunk 2 above)
Comment (apaszke-g) [reply Sources/SIL/SILParser.swift]: Hmm I didn't want to pollute the instruction with too many fields, and I don't know whether they're exclusive or not, so I'm not sure whether we can make that an enum too. This stuff is entirely undocumented and so I ignored it. I can add it though if you really want it.

File: Sources/SIL/SILParser.swift (hunk 3 above)
Comment (apaszke-g) [reply Sources/SIL/SILParser.swift]: Apparently SILGen might contain desugared types 😕

File: Sources/SIL/SILPrinter.swift (hunk 4 above)
Comment (apaszke-g) [reply Sources/SIL/SILPrinter.swift]: Oh I must have forgotten to add those. Will do.

**Comment by burmako** [comment]
Could you give an example?

File: Sources/SIL/SILParser.swift (hunk 3 above)
Comment (apaszke-g) [reply Sources/SIL/SILParser.swift]: Here you go:
```
%149 = apply %148<[Int], PartialRangeFrom<Int>>(%143, %146, %144)
```

Comment (apaszke-g) [reply Sources/SIL/SILParser.swift]: Hmmm now that I think about it, it might be better to normalize `[T]` to `Array<T>`

**Comment by burmako** [comment]
I think it would be could to update the parser to treat [T] as a special
kind of type instead of stuffing it into a name.

**Comment by apaszke-g** [comment]
Yep, I've implemented the desugaring in the parser. That was an ugly hack I made to get my stuff to work (I don't need those specializations), so thanks for catching this! I think all the comments should have been addressed by my last commit.

File: Sources/SIL/SILParser.swift [hunk 5]
```diff
@@ -610,7 +643,7 @@ class SILParser: Parser {
     func parseTypeName() throws -> String {
         let start = position
         // TODO(#14): Make name parsing more thorough.
-        let name = take(while: { $0.isLetter || $0.isNumber || $0 == "_" })
+        let name = take(while: { $0.isLetter || $0.isNumber || "_".contains($0) })
```
Comment (burmako) [inline Sources/SIL/SILParser.swift]: Let's change this back to `$0 == "_"`? This reads a bit strangely.

File: Sources/SIL/SILPrinter.swift [hunk 6]
```diff
@@ -157,12 +164,25 @@ class SILPrinter: Printer {
             print(type)
             print(", ")
             literal(value)
-        case let .load(operand):
+        case let .load(maybeOwnership, operand):
             print("load ")
+            if let ownership = maybeOwnership {
+                print(ownership)
+                print(" ")
+            }
```
Comment (burmako) [inline Sources/SIL/SILPrinter.swift]: `print(maybeOwnership, " ")`?

File: Sources/SIL/SILPrinter.swift [hunk 7]
```diff
@@ -172,10 +192,14 @@ class SILPrinter: Printer {
...
             print(operand)
-        case let .store(value, operand):
+        case let .store(value, maybeOwnership, operand):
             print("store ")
             print(value)
             print(" to ")
+            if let ownership = maybeOwnership {
+                print(ownership)
+                print(" ")
+            }
```
Comment (burmako) [inline Sources/SIL/SILPrinter.swift]: Ditto.

File: Sources/SIL/SILPrinter.swift (hunk 6 above)
Comment (apaszke-g) [reply Sources/SIL/SILPrinter.swift]: Ok, if we really want to code golf this, then I just realized that the current pattern for handling cases like this is
```
print(instructionDef.result, " = ") { print($0) }
```
which seems _really_ awkward to me. I can change it to that if you feel strongly about it though.

File: Sources/SIL/SILParser.swift (hunk 5 above)
Comment (apaszke-g) [reply Sources/SIL/SILParser.swift]: sgtm

File: Sources/SIL/SILPrinter.swift (hunk 6 above)
Comment (apaszke-g) [reply Sources/SIL/SILPrinter.swift]: We also use the `if` pattern [here (twice)](https://github.com/tensorflow/swift/blob/master/Sources/SIL/SILPrinter.swift#L322-L330) and [here](https://github.com/tensorflow/swift/blob/master/Sources/SIL/SILPrinter.swift#L135-L138).

**Pull request merged**